src/blockbroadcast.jl

```diff
@@ -105,48 +101,52 @@ julia> [idx.indices[1] for idx in SubBlockIterator(subcumulsize, cumulsize)]
 ```
 """
 struct SubBlockIterator
-    subcumulsize::Vector{Int}
-    cumulsize::Vector{Int}
+    subblock_cumsum::Vector{Int}
+    block_cumsum::Vector{Int}
 end
 
 Base.IteratorEltype(::Type{<:SubBlockIterator}) = Base.HasEltype()
 Base.eltype(::Type{<:SubBlockIterator}) = BlockIndexRange{1,Tuple{UnitRange{Int64}}}
 
 Base.IteratorSize(::Type{<:SubBlockIterator}) = Base.HasLength()
-Base.length(it::SubBlockIterator) = length(it.cumulsize) - 1
+Base.length(it::SubBlockIterator) = length(it.block_cumsum)
 
-SubBlockIterator(arr::AbstractArray, bs::BlockSizes, dim::Integer) =
-    SubBlockIterator(cumulsizes(arr, dim), cumulsizes(bs, dim))
+SubBlockIterator(arr::AbstractArray, bs::NTuple{N,AbstractUnitRange{Int}}, dim::Integer) where N =
+    SubBlockIterator(_block_cumsum(axes(arr, dim)), _block_cumsum(bs[dim]))
 
 function Base.iterate(it::SubBlockIterator, state=nothing)
     if state === nothing
-        i = 1
-        j = 1
+        i,j = 1,1
     else
         i, j = state
     end
-    length(it.cumulsize) == i && return nothing
-    idx = it.cumulsize[i]:it.cumulsize[i + 1] - 1
-    bir = BlockIndexRange(Block(j), idx .- (it.subcumulsize[j] - 1))
-    if it.subcumulsize[j + 1] == it.cumulsize[i + 1]
+    length(it.block_cumsum)+1 == i && return nothing
+    idx = i == 1 ? (1:it.block_cumsum[i]) : (it.block_cumsum[i-1]+1:it.block_cumsum[i])
+
+    bir = BlockIndexRange(Block(j), j == 1 ? idx : idx .- it.subblock_cumsum[j-1])
+    if it.subblock_cumsum[j] == it.block_cumsum[i]
         j += 1
     end
     return (bir, (i + 1, j))
 end
 
-subblocks(::Any, bs::BlockSizes, dim::Integer) =
-    (nothing for _ in 1:nblocks(bs, dim))
+subblocks(::Any, bs::NTuple{N,AbstractUnitRange{Int}}, dim::Integer) where N =
+    (nothing for _ in blockaxes(bs[dim], 1))
 
-function subblocks(arr::AbstractArray, bs::BlockSizes, dim::Integer)
+function subblocks(arr::AbstractArray, bs::NTuple{N,AbstractUnitRange{Int}}, dim::Integer) where N
     if size(arr, dim) == 1
-        return (BlockIndexRange(Block(1), 1:1) for _ in 1:nblocks(bs, dim))
+        return (BlockIndexRange(Block(1), 1:1) for _ in blockaxes(bs[dim], 1))
     end
     return SubBlockIterator(arr, bs, dim)
 end
 
 @inline _bview(arg, ::Vararg) = arg
 @inline _bview(A::AbstractArray, I...) = view(A, I...)
 
+@inline function Base.Broadcast.materialize!(dest, bc::Broadcasted{BS}) where {Style,BS<:AbstractBlockStyle}
+    return copyto!(dest, Base.Broadcast.instantiate(Base.Broadcast.Broadcasted{BS}(bc.f, bc.args, combine_blockaxes.(axes(dest),axes(bc)))))
```

Comment:
I was puzzled why you don't do this inside `copyto!` but then realized that this is because the default `materialize!` overwrites `bc.axes`:

```julia
@inline function materialize!(dest, bc::Broadcasted{Style}) where {Style}
    return copyto!(dest, instantiate(Broadcasted{Style}(bc.f, bc.args, axes(dest))))
end
```

Reply:
I might be slightly safer (and simpler?) to just ignore `bc.axes`?  Something like this

```diff
diff --git a/src/blockbroadcast.jl b/src/blockbroadcast.jl
index 91c6bb7..c97196e 100644
--- a/src/blockbroadcast.jl
+++ b/src/blockbroadcast.jl
@@ -143,10 +143,6 @@ end
 @inline _bview(arg, ::Vararg) = arg
 @inline _bview(A::AbstractArray, I...) = view(A, I...)

-@inline function Base.Broadcast.materialize!(dest, bc::Broadcasted{BS}) where {Style,BS<:AbstractBlockStyle}
-    return copyto!(dest, Base.Broadcast.instantiate(Base.Broadcast.Broadcasted{BS}(bc.f, bc.args, combine_blocka>
-end
-
 @generated function copyto!(
         dest::AbstractArray,
         bc::Broadcasted{<:AbstractBlockStyle{NDims}, <:Any, <:Any, Args},
@@ -179,7 +175,7 @@ end
     end

     quote
-        bs = axes(bc)
+        bs = map(combine_blockaxes, axes(dest), Broadcast.combine_axes(bc.args...))
         if !blockisequal(axes(dest), bs)
             copyto!(PseudoBlockArray(dest, bs), bc)
             return dest
```

Reply:
I would like to leave it as is for lazy arrays: I believe I use this so that `BroadcastArray` has the correct block structure.